src/plugins/event-plugins/FetchPlugin.ts

```diff
@@ -106,19 +103,37 @@ export class FetchPlugin extends MonkeyPatched implements Plugin {
         );
         xRayTraceEvent.subsegments.push(subsegment);
 
-        if (this.addXRayTraceIdHeader()) {
-            if (!init) {
-                init = {};
-                [].push.call(argsArray, init);
-            }
-            addAmznTraceIdHeader(
-                init,
+        if (this.config.addXRayTraceIdHeader) {
+            this.addXRayTraceIdHeader(input, init, argsArray, xRayTraceEvent);
+        }
+
+        return xRayTraceEvent;
+    };
+
+    private addXRayTraceIdHeader = (
+        input: RequestInfo,
+        init: RequestInit,
+        argsArray: IArguments,
+        xRayTraceEvent: XRayTraceEvent
+    ) => {
+        if ((input as Request).headers) {
+            return addAmznTraceIdHeaderToHeaders(
+                (input as Request).headers,
                 xRayTraceEvent.trace_id,
                 xRayTraceEvent.subsegments[0].id
             );
         }
 
-        return xRayTraceEvent;
+        if (!init) {
+            init = {};
+            [].push.call(argsArray, init);
```

Comment:
Or maybe `(argsArray as any[]).push(init)`

Reply:
Agree this would be more readable, however TypeScript doesn't like this :-)